### PR TITLE
tests: permission policy

### LINKED ISSUE #376: access: new permission policy
Implement a new permission policy, that uses the access data to restrict access to records. Will need:

- [ ] Need generators for permission levels.
- [ ] Define the permission policy similar to below (perhaps a separate ticket is needed for the handling the files permissions)
- [x] Extensive tests for the permission policy, as well as the usage of the policy inside the service.

We need two need generators:
- ``IfRestricted``
- ``RecordPermissionLevel``

Both need generators have:
- Given a record produce required needs (see (1) below)
- Given an identity produce a query filtering records to only visible records (see (2) below).

### Permission policy

The policy should look something like this:

```python
class RDMRecordPermissionPolicy(...):
    can_search = [AnyUser()]
    can_create = [AuthenticatedUser()]
    can_read = [IfRestricted('metadata', RecordPermissionLevel('viewmeta'), else_=AnyUser())] 
    can_update = [Disable()]
    can_delete = [Disable()]

    can_read_files = [IfRestricted('files', RecordPermissionLevel('viewfull'), else_=AnyUser())]
    can_updates_files = [Disable()]

    can_read_draft = [RecordPermissionLevel('edit)]
    can_update_draft = [RecordPermissionLevel('edit)]
    can_delete_draft = [RecordPermissionLevel('manage')]

    can_read_draft_files = [RecordPermissionLevel('edit')]
    can_read_update_files = [RecordPermissionLevel('edit')]
    
    can_publish = [RecordPermissionLevel('edit')]
    can_manage = [RecordPermissionLevel('manage')]
```

### 1. Record to Needs

Given a record:

```python
{
    "access": {
        "owned_by": [{"user": 1}]
        "record": "restricted",
        "files": "restricted",
        "grants": [
             {"subject": "user", "id": "2", "level": "edit"},
             {"subject": "role", "id": "curator", "level": "edit"},
             {"subject": "user", "id": "3", "level": "viewmeta"},
         ]
    }
}
```

The action ``can_read_files`` defined as ``[IfRestricted('files', RecordPermissionLevel('viewfull'), else_=AnyUser())]`` should require the following needs:

```
# delegate to record.access.needs('viewfull') to make the needs:
UserNeed(1), UserNeed(2), RoleNeed("curator")
```

Note:
- Owners are included because they have implicit "manage" permission.
- User 2 and role curator both have edit level so they have implicit viewfull-level
- User 3 only has viewmeta level so the user is not included.
- All of above should be done and working in the system field (#372)

### 2. Identity to Query

The query would take an identity providing something like (note we may need special handling of any user/authenticated user to handle it efficiently:

```
UserNeed(1), RoleNeed('curator'), SystemRole('authenticated_user'), SystemRole('any_user')
```

and make a query similar to:

```
access.record:public 
OR (
    access.record:restricted AND (
        access.tokens:view-user-1 OR 
        access.tokens:view-role-curator OR          
        access.tokens:view-sysrole-authenticated_user OR
        access.tokens:view-sysrole-any_user
    )
)
```

### PR DESCRIPTION
**Pull request by mb-wali**
related to #376 

### Permission policy

```python
class TestRDMPermissionPolicy(....):
    """Define permission policies for RDM Records."""
    can_search = [AnyUser(), SystemProcess()]
    can_create = [AuthenticatedUser(), SystemProcess()]
    can_update = [Disable()]
    can_delete = [Disable()]
    can_read = [
        IfRestricted('record',
                     then_=[RecordOwners()],
                     else_=[AnyUser(), SystemProcess()])
                     ]
    can_read_files = [
        IfRestricted('files',
                     then_=[RecordOwners()],
                     else_=[AnyUser(), SystemProcess()])
                     ]
    can_update_files = [Disable()]
    can_read_draft = [RecordOwners()]
    can_update_draft = [RecordOwners()]
    can_delete_draft = [RecordOwners()]
    can_read_draft_files = [RecordOwners()]
    can_read_update_files = [RecordOwners()]
    can_publish = [RecordOwners()]
    can_manage = [RecordOwners()]
```
----

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -301,3 +305,46 @@ def identity_simple():
...
     i.provides.add(Need(method='system_role', value='any_user'))
     return i
+
+
+@pytest.fixture(scope="module")
+def identity_authenticated():
+    """Simple identity fixture."""
+    i = Identity(1)
+    i.provides.add(UserNeed(1))
+    i.provides.add(Need(method='system_role', value='authenticated_user'))
```
Comment (lnielsen) [inline tests/conftest.py:315]: from invenio_access.permissions import authenticated_user
authenticated_user

File: tests/conftest.py [hunk 2]
```diff
@@ -324,3 +329,45 @@ def identity_simple():
...
+
+@pytest.fixture(scope="module")
+def identity_authenticated():
+    """Simple identity fixture."""
+    identity = Identity(1)
+    identity.provides.add(authenticated_user)
+    return identity
+
+
+@pytest.fixture()
```
Comment (lnielsen) [inline tests/conftest.py:342]: Minor: Can you check if it's possible to use [these functions](https://github.com/inveniosoftware/invenio-accounts/blob/master/invenio_accounts/testutils.py) from Invenio-Accounts instead?

**Comment by fenekku** [comment]
This looks good to me, however with the recent merges into invenio-rdm-records, there might need some adaptation. I've attempted a rebase, but with the 14 commits here, that isn't easy :smile: . 

I *think* the simplest for you @mb-wali would be to extract the 2 modified files, then sync your local master and start a new branch from it where you add back the extracted modified files (you may git reset --hard this branch locally to master and add the 2 modified files on top to keep the same branch / discussion flow on GitHub). From there, you should see what fails or not and be able to make the changes.

I hope this is relatively clear / it's a Friday :smile: .

**Review by fenekku — APPROVED** [review]

**Comment by mb-wali** [comment]
> This looks good to me, however with the recent merges into invenio-rdm-records, there might need some adaptation. I've attempted a rebase, but with the 14 commits here, that isn't easy  .
> 
> I _think_ the simplest for you @mb-wali would be to extract the 2 modified files, then sync your local master and start a new branch from it where you add back the extracted modified files (you may git reset --hard this branch locally to master and add the 2 modified files on top to keep the same branch / discussion flow on GitHub). From there, you should see what fails or not and be able to make the changes.
> 
> I hope this is relatively clear / it's a Friday  .

done

**Review by fenekku — APPROVED** [review]

**Pull request merged**